### ISSUE TITLE
Error while saving composite entity having another entity id as null

### ISSUE THREAD

**Issue by vtiwari2134** [javers/javers#712]
Following exception occurs while saving entity with javers auditing enabled in spring - 
JaversException ENTITY_INSTANCE_WITH_NULL_ID: Found Entity instance 'com.test.entity.DepartmentEntity' with null idProperty 'id'

I think the issue is with <code>JaversSpringDataAuditableRepositoryAspect.class</code> having aspect on save (<code>@AfterReturning("execution(public * save(..)) && this(org.springframework.data.repository.CrudRepository)")</code>) method. In this the entity is audited on the argument passed to save method but auditing should be performed on return entity that is saved by hibernate.

Issue arises because whenever detached entity is saved to persist hibernate tries to merge the entity creating new entity in hibernate session. Hence the return entity is different from the entity passed to save.
That why the Aspect <code>public void onSaveExecuted(JoinPoint pjp) {
    this.onVersionEvent(pjp, this.saveHandler);
  }</code> should evaluate the return entity from <code>save</code> method.

**Sample spring project to replicate bug -**
[test.zip](https://github.com/javers/javers/files/2385977/test.zip)

To replicate issue run JaversExceptionTest  test case. This test class contains 2 test one saves entity without having auditing enabled and one with javers auditing enabled.
In first case EmployeeEntity is saved successfully in db while second case throws error : -

<code>JaversException ENTITY_INSTANCE_WITH_NULL_ID: Found Entity instance 'com.test.entity.DepartmentEntity' with null idProperty 'id'

	at org.javers.core.metamodel.type.EntityType.getIdOf(EntityType.java:88)
	at org.javers.core.metamodel.object.InstanceId.createFromInstance(InstanceId.java:28)
	at org.javers.core.metamodel.object.GlobalIdFactory.createId(GlobalIdFactory.java:49)
	at org.javers.core.graph.LiveCdoFactory.create(LiveCdoFactory.java:25)
	at org.javers.core.graph.LiveCdoFactory.create(LiveCdoFactory.java:10)
	at org.javers.core.graph.EdgeBuilder.buildSingleEdge(EdgeBuilder.java:30)</code>

**Comment by bartoszwalacik** [comment]
@vtiwari2134 could you please push this test case to github? see https://github.com/javers/javers/tree/master/javers-spring-boot-starter-sql/src/test/groovy/org/javers/spring/sql

**Comment by vtiwari2134** [comment]
> @vtiwari2134 could you please push this test case to github? see https://github.com/javers/javers/tree/master/javers-spring-boot-starter-sql/src/test/groovy/org/javers/spring/sql
Sure, I will push the test cases. Do you want me to push it to above repository or any new repository would work ?

> btw, in your test, employeeEntity is a new object not detached one (before hibernate generates Id for it), right?
Yes employeeEntity is a new object for which i am generating a random id and try to save it.

And thanks for quick reply.

**Comment by bartoszwalacik** [comment]
Pull request from your fork of javers repo to this repo would be perfect. Then, I only have to switch to your branch.

**Comment by vtiwari2134** [comment]
Created pull request - https://github.com/javers/javers/pull/713 for this issue.

**Comment by bartoszwalacik** [comment]
thnx

**Comment by bartoszwalacik** [comment]
ok, so the problem is that you have different mapping style for hibernate (fields) and for JaVers (beans).
I have modified your PR to provide better isolation of this issue, see https://github.com/javers/javers/pull/716/files

**Comment by bartoszwalacik** [comment]
btw, all pointcuts are defined `@AfterReturning`, see https://github.com/javers/javers/blob/master/javers-spring-jpa/src/main/java/org/javers/spring/auditable/aspect/springdatajpa/JaversSpringDataJpaAuditableRepositoryAspect.java

**Comment by bartoszwalacik** [comment]
After some digging, mapping style is not a problem. I dont konw the reason yet.

**Comment by vtiwari2134** [comment]
I triaged a bit and this is what i think -
When we pass entity with id generated manually with save method of JPARepository, it checks whether the entity is new or not.
```
public <S extends T> S save(S entity) { 
   /*** Check if id is null then persist else merge it */
    if (this.entityInformation.isNew(entity)) {  
      this.em.persist(entity);
      return entity;
    } else {
      return this.em.merge(entity);
    }
  }
```
Since we are generating id it will try to merge it as per the above code. Now the above entity it will treat as detached from session so it will create new entity in session and will save that with the ids.
So, the entity that we sent in argument for save would be different from the return object.

That why the Aspect `public void onSaveExecuted(JoinPoint pjp) { this.onVersionEvent(pjp, this.saveHandler); }` should evaluate the return entity from save method not the one that is passed in pjp argument list.

It should more like -
```
@AfterReturning("execution(public * save(..)) && this(org.springframework.data.repository.CrudRepository)", returning="returnObject")
  public void onSaveExecuted(JoinPoint pjp, Object returnObject) {
   // Check should be performed on returnObject
    this.onVersionEvent(pjp, this.saveHandler);
  }
```

**Comment by bartoszwalacik** [comment]
Thanks for the insight! Would you like to create second PR with the fix?

**Comment by vtiwari2134** [comment]
Sure i will create a PR with the fix.

**Comment by vtiwari2134** [comment]
PR - https://github.com/javers/javers/pull/718

**Comment by bartoszwalacik** [comment]
ok, but tests are failing :) looks like commit in't called now

**Comment by vtiwari2134** [comment]
Sorry, I just tested test cases for employee entity object. I will check this tomorrow, busy with some work. :)

**Comment by vtiwari2134** [comment]
@bartoszwalacik 
Hi, I have made changes to the aspect, fix all the test cases and raised PR.
PR - https://github.com/javers/javers/pull/719
Please have a look whether it makes sense.

I have added inline description in test-cases where i felt it is required.
Do tell if more description is required for the changes.

**Comment by bartoszwalacik** [comment]
ok, thanks, I will do the code review

**Comment by bartoszwalacik** [comment]
@vtiwari2134 thanks for your excellent contribution, release is on it's way to Central

**Comment by vtiwari2134** [comment]
@bartoszwalacik Thats great! Happy to contribute. :) Thanks!

**Comment by bartoszwalacik** [comment]
released in 3.11.6
